simulations/vip-171/vip-171-testnet/simulations.ts

```diff
@@ -0,0 +1,168 @@
+      inputs: [
+        {
+          indexed: false,
+          internalType: "address",
+          name: "oldImplementation",
+          type: "address",
+        },
+        {
+          indexed: false,
+          internalType: "address",
+          name: "newImplementation",
+          type: "address",
+        },
+      ],
+      name: "NewImplementation",
+      type: "event",
+    },
+  ];
+  testVip("VIP-171 Core  VToken Upgrade of AIA", vip171Testnet(), {
+    callbackAfterExecution: async txResponse => {
+      await expectEvents(
+        txResponse,
+        [VTOKEN_ABI, ProxyAdminInterface],
+        ["NewImplementation", "NewProtocolShareReserve", "NewReduceReservesBlockDelta", "NewAccessControlManager"],
+        [11, 11, 11, 11],
+      );
+    },
+  });
+});
+
+forking(34259644, () => {
+  describe("Post VIP simulations", async () => {
+    before(async () => {
+      await pretendExecutingVip(vip171Testnet());
+      [user] = await ethers.getSigners();
+    });
+
+    for (const market of CORE_MARKETS) {
+      it(`Save post VIP storage snapshot of ${market.name}`, async () => {
+        if (!market.isMock) {
+          user = await initMainnetUser(TOKEN_HOLDER, ethers.utils.parseEther("5"));
+        } else {
+          await setBalance(user.address, ethers.utils.parseEther("5"));
+        }
+
+        vToken = new ethers.Contract(market.address, VTOKEN_ABI, provider);
+        const comptroller = new ethers.Contract(await vToken.comptroller(), COMPTROLLER_ABI, provider);
+        underlying = new ethers.Contract(await vToken.underlying(), MOCK_TOKEN_ABI, provider);
+
+        await comptroller.connect(impersonatedTimelock)._setMarketBorrowCaps([market.address], [parseUnits("2", 48)]);
+        await comptroller.connect(impersonatedTimelock)._setMarketSupplyCaps([market.address], [parseUnits("2", 48)]);
+        await comptroller.connect(impersonatedTimelock)._setCollateralFactor(market.address, parseUnits("0.9", 18));
+
+        await performVTokenBasicActions(
+          market.address,
+          user,
+          mintAmount,
+          borrowAmount,
+          repayAmount,
+          redeemAmount,
+          vToken,
+          underlying,
+          market.isMock,
+        );
+        const state = await fetchVTokenStorageCore(vToken, user.address);
+
+        delete state.totalReserves;
+        delete state.pendingOwner;
+        delete state.owner;
+
+        expect(await vToken.implementation()).equals(NEW_VBEP20_DELEGATE_IMPL);
+        expect(await vToken.protocolShareReserve()).equals(PROTOCOL_SHARE_RESERVE);
+        expect(await vToken.accessControlManager()).equals(ACCESS_CONTROL_MANAGER);
+        expect(await vToken.admin()).equals(NORMAL_TIMELOCK);
+
+        postVipStorage.push(state);
+      });
+    }
+
+    it("Should match pre and post storage", async () => {
+      for (let i = 0; i < preVipStorage.length; i++) {
+        expect(preVipStorage[i]).to.deep.equal(postVipStorage[i]);
+      }
+    });
```

Comment:
Would it make sense to add a test doing this, for each market?
1. store the reserves of the market, and the funds in the PSR contract for the underlying token
2. mine `reduceReservesBlockDelta - 1` blocks
3. invoke `accrueInterest()` on the vToken
4. check the reserves are greater than the stored one in the market
5. mine 1 block
6. invoke `accrueInterest()` on the vToken
7. check the reserves are 0 in the market, and are the expected ones in the PSR, for this underlying

Reply:
Added here https://github.com/VenusProtocol/vips/pull/67/commits/46fef5dbf378dad1eacb8b0b0d7c83e06e47704a#diff-8fa5a44e8a45eab9177db5431042b8303568c590d759b67d1114b9824d097892R167

Block mining is not needed because it will reduce reserves in the first operation of accruing interest after upgrades.The initial value of ReservesBlockNumber is 0, and the condition is (currrentBlock - reduceReservesBlockNumber > reduceReservesBlockDelta).